lib/Doctrine/ORM/Mapping/AssociationMapping.php

```diff
@@ -0,0 +1,240 @@
+
+    public bool|null $unique = null;
+
+    /**
+     * @param string       $fieldName    The name of the field in the entity
+     *                                   the association is mapped to.
+     * @param class-string $sourceEntity The class name of the source entity.
+     *                                   In the case of to-many associations
+     *                                   initially present in mapped
+     *                                   superclasses, the nearest
+     *                                   <em>entity</em> subclasses will be
+     *                                   considered the respective source
+     *                                   entities.
+     * @param class-string $targetEntity The class name of the target entity.
+     *                                   If it is fully-qualified it is used as
+     *                                   is. If it is a simple, unqualified
+     *                                   class name the namespace is assumed to
+     *                                   be the same as the namespace of the
+     *                                   source entity.
+     */
+    final public function __construct(
+        public string $fieldName,
+        public string $sourceEntity,
+        public string $targetEntity,
+    ) {
+    }
+
+    /**
+     * @psalm-param array{
+     *     fieldName: string,
+     *     sourceEntity: class-string,
+     *     targetEntity: class-string,
+     *     joinTable?: mixed[]|null,
+     *     isOwningSide: bool, ...} $mappingArray
+     */
+    public static function fromMappingArray(array $mappingArray): static
+    {
+        unset($mappingArray['isOwningSide']);
+        $mapping = new static(
+            $mappingArray['fieldName'],
+            $mappingArray['sourceEntity'],
+            $mappingArray['targetEntity'],
+        );
+        foreach ($mappingArray as $key => $value) {
+            if (in_array($key, ['type', 'fieldName', 'sourceEntity', 'targetEntity'])) {
+                continue;
+            }
+
+            if ($key === 'joinTable') {
+                assert($mapping instanceof ManyToManyAssociationMapping);
+
+                if ($value === [] || $value === null) {
+                    continue;
+                }
+
+                assert($mapping instanceof ManyToManyOwningSideMapping);
+
+                $mapping->joinTable = JoinTableMapping::fromMappingArray($value);
+
+                continue;
+            }
+
+            if (property_exists($mapping, $key)) {
+                $mapping->$key = $value;
+            } else {
+                throw new Exception('Unknown property ' . $key . ' on class ' . static::class);
+            }
+        }
+
+        return $mapping;
+    }
+
+    /** @psalm-assert-if-true AssociationOwningSideMapping $this */
+    public function isOwningSide(): bool
+    {
+        return $this instanceof AssociationOwningSideMapping;
+    }
```

Comment:
Maybe I'm getting old, but wouldn't the saner approach be…

```suggestion
    public function isOwningSide(): bool
    {
        return false;
    }
```

… and overriding this method in `AssociationOwningSideMapping`?

Reply:
It doesn't look saner to me. When stumbling upon the code in your proposition, one has might think "what's going on here? Ah, maybe that method is overridden somewhere (but where?). With the approach I picked, you have everything you need to understand this in the same place.

Reply:
> Ah, maybe that method is overridden somewhere

Yeah, that's the general idea of abstract classes. 😅 

It's a bit weird that the abstraction knows details about its implementations. But all right, let's roll with it for now.

Reply:
> It's a bit weird that the abstraction knows details about its implementations

I'd like to make this a bit more like a sealed class (https://wiki.php.net/rfc/sealed_classes), but I didn't know how. But yeah, that's indeed the concept here.